### PR TITLE
[WIP] adding common knative controller options.

### PR DESCRIPTION
**Pull request by n3wscott**
Trying to remove some of the boilerplate from setting up a controller. Starting with options.

Here is an example of this change integrated with a controller: https://github.com/n3wscott/autotrigger/pull/1

### CODE REVIEW

File: controller/options.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	ResyncPeriod time.Duration
+
+	// TrackerMultiple a multiple of the resync period to use.
+	TrackerMultiple int
+
+	// StopChannel is the shared stop channel to end the process.
+	StopChannel <-chan struct{}
+
+	// TODO: We should have a common stats reporter, but these are custom at the moment.
+	//StatsReporter StatsReporter
```
Comment (mattmoor) [inline controller/options.go]: cc @mdemirhan

File: controller/options.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		KubeClientSet:    kubeClient,
+		SharedClientSet:  sharedClient,
+		DynamicClientSet: dynamicClient,
+		Logger:           logger,
+		ResyncPeriod:     10 * time.Hour, // Based on controller-runtime default.
+		TrackerMultiple:  3,              // Based on knative usage.
+		StopChannel:      stopCh,
+	}
+
+	return opts
```
Comment (mattmoor) [inline controller/options.go]: `return Options{`

File: controller/options.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+// refresh leases frequently enough that we don't miss updates to tracked
+// objects.
+func (o Options) GetTrackerLease() time.Duration {
+	return o.ResyncPeriod * time.Duration(o.TrackerMultiple)
+}
+
+func NewOptions(ctx context.Context, cfg *rest.Config, stopCh <-chan struct{}) Options {
+	logger := logging.FromContext(ctx)
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
```
Comment (mattmoor) [inline controller/options.go]: We just changed all of these to `NewForConfigOrDie` FWIW.

File: controller/options.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		logger.Fatalf("Version check failed: %v", err)
+	}
+
+	opts := Options{
+		KubeClientSet:    kubeClient,
+		SharedClientSet:  sharedClient,
+		DynamicClientSet: dynamicClient,
+		Logger:           logger,
+		ResyncPeriod:     10 * time.Hour, // Based on controller-runtime default.
+		TrackerMultiple:  3,              // Based on knative usage.
```
Comment (mattmoor) [inline controller/options.go]: hoist these as `const`?

File: controller/options.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalw("Error building dynamic clientset", zap.Error(err))
+	}
+
+	if err := version.CheckMinimumVersion(kubeClient.Discovery()); err != nil {
+		logger.Fatalf("Version check failed: %v", err)
+	}
```
Comment (mattmoor) [inline controller/options.go]: We also added the scale client to our common options recently.  Makes it easier to write autoscaler plugins to have it here 🤷‍♂️

Comment (n3wscott) [reply controller/options.go]: I am suggesting you have three levels: pkg, project base, specific controller options if needed.

**Comment by n3wscott** [comment]
/hold 

I am going to rework this a little.

**Comment by n3wscott** [comment]
/hold cancel

RFAL

File: controller/options.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	MetricsConfigPath string
+	MetricsConfigName string
+	MetricsObserver   metrics.ObserverDecorator
+}
+
+// NewOptions creates the common to Knative controller options.
+// component is the name of the controller component.
+// loggingConfigFile is the file path to the logging config map.
+func NewOptions(component string, cfg *rest.Config, configCfg ConfigMapConfig) Options {
+	loggingConfigMap, err := configmap.Load(configCfg.LoggingConfigPath)
```
Comment (n3wscott) [inline controller/options.go]: I now make loggers here.

File: controller/options.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatalf("Error parsing logging configuration: %v", err)
+	}
+	logger, atomicLevel := logging.NewLoggerFromConfig(loggingConfig, component)
+	defer logger.Sync()
+
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
```
Comment (n3wscott) [inline controller/options.go]: And the stop channel.

File: controller/options.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		Logger:           logger,
+		AtomicLevel:      atomicLevel,
+		ResyncPeriod:     defaultResyncPeriod,
+		TrackerMultiple:  defaultTrackerMultiple,
+		StopChannel:      stopCh,
+	}
+	opts.ConfigMapWatcher = configmap.NewInformedWatcher(opts.KubeClientSet, system.Namespace())
+
+	// Watch the logging config map and dynamically update logging levels.
+	opts.ConfigMapWatcher.Watch(configCfg.LoggingConfigPath, configCfg.LoggingObserver(logger, atomicLevel))
```
Comment (n3wscott) [inline controller/options.go]: And watch for config changes.

File: controller/options.go (hunk 3 above)
Comment (n3wscott) [reply controller/options.go]: now using NewForConfigOrDie

File: controller/options.go (hunk 1 above)
Comment (mdemirhan) [reply controller/options.go]: Responding about ... 30 days late :)

I think it is worth thinking about creating a generalized StatsReporter. I thought about it before, but ultimately failed to do something that would be simpler than what OpenCensus provides. Let me know if you have any ideas around making the current situation better.

**Pull request closed without merging**